### DISCUSSION TITLE
Kraken Z73 | set lcd screen gif | Assertion Error ('Cannot find bulk out device')

### DISCUSSION THREAD

**Discussion by CriticalPoint** [liquidctl/liquidctl#621] [Q&A]
Hey folks 👋🏻

Until recently I had this working well so I suspect that my issue is more related to Python than it is LiquidCTL!

In a bid to bring my python knowledge from basically none to at least being able to understand how it pulls together, and maybe a bit of syntax, I've ended-up basically bricking my python by having loads of different versions - I found that PATH was confusingly pointing at InkScape! 🤷🏻‍♂️

Anyway, I now have fresh Python [3.11.4], a `venv` with `liquidctl` and I'm able to set pump duty, fan speed and brightness on the Kraken Z73, but what I have not been able to do is set a GIF (or static image) due to this error:

> ERROR: NZXT Kraken Z (Z53, Z63 or Z73) (experimental): unexpected error: AssertionError('Cannot find bulk out device')

I've followed all the guides and also updated the `LibUSB 1.0 DLL`, since the python version changed - and I've been scrolling through the code line by line for around 5 hours now, but cannot make sense of why this just isn't working and why I am able to _partially_ access the device..

The code surrounding the error:
![cannot find bulk out device](https://github.com/liquidctl/liquidctl/assets/3483630/38a87f06-d872-4720-8336-1b743242e492)

I had the suspicion that it could be related to file size even though it worked before and so, assuming nothing, I tried with some smaller files and I hit the same error every time. I checked this as that specific line of code flags in the `--verbose` error (which is in the screenshot at the bottom).

Here's the code surrounding the file size. Seems unrelated.
![FileSize](https://github.com/liquidctl/liquidctl/assets/3483630/d6ae8c82-2e32-42ca-94d4-a890c7d66628)

My most recent thought was that the data is loaded into a bucket on a separate interface, which is why I can do some things and not others, so I checked with Zadiag and I see that there ARE two, and that they ARE configured differently...

`Interface 0`
![Zadiag Interface 0](https://github.com/liquidctl/liquidctl/assets/3483630/9bae1940-dc18-4e3d-8351-20bd1a755285)
`Interface 1`
![Zadiag Interface 1](https://github.com/liquidctl/liquidctl/assets/3483630/96483cb6-8050-4e78-b6c4-1939c9703ebe)

So what I'm thinking is that I press this comedy-sized button
![Zadiag Interface 1_p](https://github.com/liquidctl/liquidctl/assets/3483630/c4085c42-1676-4d87-b51f-ec62c4d5ebbf)
and replace `HidUsb` with `WinUSB`, and that fixes it - but I worry that in actual fact I'll just lose access altogether.

> And that's why I'm here - to ask you knowledgeable peeps if it's really that simple!?

Here's the process I've been through which might shed some light on this - if anyone can see what stoooopid thing I have done (PATH, DIR or syntax issues) then please, just put me out of my misery! 🙏🏻

![LiquidCTL GIF Verbose Log](https://github.com/liquidctl/liquidctl/assets/3483630/06a64193-f694-49bf-8c16-ce9b09913fb7)

**Comment by aleksamagicka** [comment]
> and replace HidUsb with WinUSB, and that fixes it - but I worry that in actual fact I'll just lose access altogether.

Just tried this on a fresh VM with a new Python install and venv. I replaced the driver in Zadig and liquidctl does not see my Kraken Z53 anymore. In Zadig I see the same drivers for the two endpoints as you, so there is probably something else at play.

**Reply by CriticalPoint** [reply]
I should have thought to do that in VirtualBox 🤦🏻‍♂️ You're a hero for trying this.

OK - So the solution seems to be, as I feared, buried behind a bunch of Python debugging, which is not my strong suite!

I think I will do as you did and spin up a VM to debug without fear of doing something permanent in my OS. I reckon I have a couple hours in me over the weekend, but if it works right away in the VM, I will almost certainly end up blitzing my machine and starting over! 😅

Thank you for your effort and your energy 😇 🙏🏻

**Reply by CriticalPoint** [reply]
Call me a martyr, but I may actually just go a step further and write this out in PS to deploy automatically to Windows Sandbox; that way I can work on this over time and at the end of it I'll have a script to configure everything for me 🤓

**Reply by CriticalPoint** [reply]
Did you switch the driver back `WinUSB (v6.1.7600.16385) --> HidUsb (vl0.0.19041.3031)` and if so, did liquidctl regain sight of it, or did you just bin off the VM? 🙂

**Comment by aleksamagicka** [comment]
I didn't notice that it could be reversed back to HidUsb? I'll look
tomorrow.

**Reply by aleksamagicka** [reply]
Zadig doesn't allow reverting back to HidUsb. The suggested solution, as per their issue tracker, is to uninstall the devices from the Device Manager and reboot. Didn't work for me in a VM. In all likeliness the problem that you're encountering is coming from somewhere else, and not the driver Windows assigned it.

**Comment by CriticalPoint** [comment]
Lots of stuff's happened to get this sorted! Short version is that I uninstalled the device using USBDeview, but Windows just kept assigning the same driver back and so I ended up disconnecting the Kraken and scanned through the registry (exporting and deleting) any and all mention of _NZXT_, _VID_1E71_, and/or _PID_3008_.

I then went in to find the drivers (found in the reg entries), `"DeviceDesc"="@oem180.inf,%devicename%;NZXT KrakenZ Device"`, also the `ClassGUID's`, and added those to a `REG DELETE /F` batch file, which I then loged-out and ran in start-up repair. Seemed to get most of them. I keep a log of everything that got ripped out, which ended up being quite a lot (10 or so registry entries).

I then installed CAM (for drivers) and after a couple reboots, I plugged the Z73 in and rebooted.
Good news - I could now see it with LiquidCTL, and can set the brightness. Progress!
Bad news - not even CAM can see the Kraken, and I cannot load GIFs 😭

![image](https://github.com/liquidctl/liquidctl/assets/3483630/918d2841-5814-4083-a20d-f20a0bf8f19d)

I am now starting to think that the 'main driver' for the device isn't the same as the driver that accesses storage - more work to do here to get to the bottom of that [_goes to check USBDeview..._]

![image](https://github.com/liquidctl/liquidctl/assets/3483630/a32b035e-b34d-4e6d-949c-a8ead714496d)

Yeah, "_Device Name : 0000.0014.0000.013.001.001.000.000.000_" looks like a promising place to start 🤩 I might not know what I am doing, but I won't give up trying - the worst kind of tinkerer 😅 😂

It's either this, or I've managed to some how completely bork the storage on the Kraken, possibly by uploading too much data, or causing overflow (I recall reading something about buckets...). No clue on that, but that's my final thought if this 'other driver' doesn't resolve it.

Thanks for listening - update to follow 🤷🏻‍♂️

**Comment by CriticalPoint** [comment]
Holy sheet, CAM can now see the device 🤯

![image](https://github.com/liquidctl/liquidctl/assets/3483630/a85d6b2d-7fa1-4eb3-ba6f-f8eaece9c84b)

It's not ideal, but it is progress!

**Comment by CriticalPoint** [comment]
I've tried recreating the `venv` and calling `pip install liquidctl` with the following versions;
- Python 3.8.10
- Python 3.10.9
- Python 3.11.5

I verified each with `python -m pytest`, and I am still getting the same error each time - thus it doesn't appear to be related to the Python version.

Confirmed that each time I am placing `libusb-1.0.dll` into the `Scripts` folder within the `venv`. No dice.

I'm tired and done for the evening (errrrm, morning 🥱), so I will work out a plan to partition off some SSD and blat Windows onto it to tinker in a completely new and isolated environment as I really do now want to see this through.

**Reply by jb901** [reply]
Don't suppose you ever made any headway on this?  Having the same problem with my Elite 360 2024 and put myself in the same boat as you with the driver :(

**Reply by btcthoty** [reply]
same here, any solutions would be really appreciated.
